### PR TITLE
Agent streaming

### PR DESCRIPTION
**Pull request by bracesproul**

### CODE REVIEW

File: langchain/src/agents/agent_iterator.ts [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+  metadata?: Record<string, unknown>;
+  runName?: string;
+  runManager?: CallbackManagerForChainRun;
+}
+
+export class AgentExecutorIterator
+  extends Serializable
+  implements AgentExecutorIteratorInput
+{
+  lc_namespace = ["langchain", "agents", "executor", "iterator"];
```
Comment (jacoblee93) [inline langchain/src/agents/agent_iterator.ts:25]: This is the correct namespace?

File: langchain/src/agents/agent_iterator.ts [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+  get nameToToolMap(): Record<string, Tool> {
+    const toolMap = this.agentExecutor.tools.map((tool) => ({
+      [tool.name]: tool,
+    }));
+    return Object.assign({}, ...toolMap);
+  }
+
+  constructor(fields: AgentExecutorIteratorInput) {
+    super();
```
Comment (jacoblee93) [inline langchain/src/agents/agent_iterator.ts:69]: Pass `fields` into `super` for serialization

File: langchain/src/agents/agent_iterator.ts [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+
+  /**
+   * Increment the number of iterations and update the time elapsed.
+   */
+  updateIterations(): void {
+    this.iterations += 1;
+  }
+
+  /** Method to initialize the iterator */
+  async *[Symbol.asyncIterator]() {
```
Comment (jacoblee93) [inline langchain/src/agents/agent_iterator.ts:96]: We sure we need this? Might be nicer to have a method that returns an `IterableReadableStream`?

File: langchain/src/agents/executor.ts [hunk 4]
```diff
@@ -299,6 +337,149 @@ export class AgentExecutor extends BaseChain<ChainValues, AgentExecutorOutput> {
...
+  ): Promise<AgentFinish | AgentStep[]> {
+    let output;
+    try {
+      output = await this.agent.plan(
+        intermediateSteps,
+        inputs,
+        runManager?.getChild()
+      );
+    } catch (e) {
+      // handle rrr
```
Comment (jacoblee93) [inline langchain/src/agents/executor.ts:354]: rrr

File: langchain/src/agents/executor.ts [hunk 5]
```diff
@@ -299,6 +337,149 @@ export class AgentExecutor extends BaseChain<ChainValues, AgentExecutorOutput> {
     return getOutput(finish);
   }
 
+  async _takeNextStep(
```
Comment (jacoblee93) [inline langchain/src/agents/executor.ts:340]: These are all only used in the streaming flow right?

File: langchain/src/chains/base.ts [hunk 6]
```diff
@@ -83,6 +83,32 @@ export abstract class BaseChain<
     return this.call(input, config);
   }
 
+  private _validateOutputs(outputs: Record<string, unknown>): void {
```
Comment (jacoblee93) [inline langchain/src/chains/base.ts:86]: These changes don't seem related? Or were we missing them all along?

File: langchain/src/agents/tests/agent.int.test.ts [hunk 7]
```diff
@@ -315,14 +315,42 @@ test("Run tool web-browser", async () => {
...
-    },
-    "Run tool web-browser"
-  );
   expect(result.intermediateSteps.length).toBeGreaterThanOrEqual(1);
   expect(result.intermediateSteps[0].action.tool).toEqual("web-browser");
   expect(result.output).not.toEqual("");
   expect(result.output).not.toEqual("Agent stopped due to max iterations.");
 });
+
+test.only("Agent can stream", async () => {
```
Comment (jacoblee93) [inline langchain/src/agents/tests/agent.int.test.ts:324]: Don't forget to remove `.only`

**Review by jacoblee93 — COMMENTED** [review]
Looks reasonable overall, see comments. Also might need to rebase with core refactor changes

File: langchain/src/agents/agent_iterator.ts (hunk 1 above)
Comment (bracesproul) [reply langchain/src/agents/agent_iterator.ts]: Maybe this?

```typescript
lc_namespace = ["langchain", "agents", "executor_iterator"];
```

File: langchain/src/agents/agent_iterator.ts (hunk 3 above)
Comment (bracesproul) [reply langchain/src/agents/agent_iterator.ts]: This is cleaner imo. But, one could make the case that calling a method you invoke would provide a better DX

File: langchain/src/agents/executor.ts (hunk 5 above)
Comment (bracesproul) [reply langchain/src/agents/executor.ts:611]: Correct

File: langchain/src/chains/base.ts (hunk 6 above)
Comment (bracesproul) [reply langchain/src/chains/base.ts:86]: Ported over from python so I'd assume we just were missing them to start with

File: langchain/src/agents/executor.ts [hunk 8]
```diff
@@ -8,18 +8,296 @@ import { StoppingMethod } from "./types.js";
...
+  }
+
+  /**
+   * Reset the iterator to its initial state, clearing intermediate steps,
+   * iterations, and time elapsed.
+   */
+  reset(): void {
+    this.intermediateSteps = [];
+    this.iterations = 0;
+    this._finalOutputs = undefined;
```
Comment (bracesproul) [inline langchain/src/agents/executor.ts:103]:
```suggestion
   * Reset the iterator to its initial state, clearing intermediate steps,
   * iterations, and the final output.
   */
  reset(): void {
    this.intermediateSteps = [];
    this.iterations = 0;
    this._finalOutputs = undefined;
```

File: langchain/src/agents/executor.ts [hunk 9]
```diff
@@ -8,18 +8,296 @@ import { StoppingMethod } from "./types.js";
...
+  }
+
+  /**
+   * Increment the number of iterations and update the time elapsed.
+   */
+  updateIterations(): void {
+    this.iterations += 1;
+  }
```
Comment (bracesproul) [inline langchain/src/agents/executor.ts:111]:
```suggestion

  updateIterations(): void {
    this.iterations += 1;
  }
```

File: langchain/src/agents/tests/agent.int.test.ts [hunk 10]
```diff
@@ -322,7 +322,58 @@ test("Run tool web-browser", async () => {
     "Run tool web-browser"
   );
   expect(result.intermediateSteps.length).toBeGreaterThanOrEqual(1);
-  expect(result.intermediateSteps[0].action.tool).toEqual("web-browser");
+  expect(result.intermediateSteps[0].action.tool).toEqual("search");
+  expect(result.intermediateSteps[1].action.tool).toEqual("web-browser");
```
Comment (bracesproul) [inline langchain/src/agents/tests/agent.int.test.ts:326]: Was failing even on a fresh branch
